### DISCUSSION TITLE
フォロー・フォロワー機能を悪用したスパムへの対策

### DISCUSSION THREAD

**Discussion by ko1nksm** [increments/qiita-discussions#76] [要望・アイデア・不具合]
最近というか人によって異なると思うのですが、フォロー・フォロワー機能を悪用したスパムが酷いです。

マイページ（ユーザーページ）にある「フォロワー」の一覧は、自分もしくは他の人が興味があって見るためのページとして作られていると思いますが、そのページが広告で埋め尽くされることになっており、また通知機能はスパム通知機能となっています。（通知を無効にできることは確認済みです。）

おそらくQiitaトップページのユーザーランキングに掲載されてからだと思うので、Qiiaのユーザーランキング上位者ほどスパムに悩まされる事態になっていると考えられます。ユーザーランキングに掲載されている人を対象に同じスパム業者がフォローしているのを確認できます。

何かしらの対策をお願いしたいです。

例えば「記事投稿数が0のユーザーは（フォロー機能は使えるものの）他の人のフォロワーリストに表示されない（通知もされない）」というのは受け入れられるのではないでしょうか？スパム業者に限らず記事を何も投稿してないのであればその人に興味を持つ人はいないはずで、フォロワー一覧に表示する理由もないと思われます。ダミー記事を投稿することで回避可能ですが、スパム業者としてはスパム記事を書く方が効果が高いはずですので、それを行っていないところを見るとスパム記事への対策はすでに十分行われていると考えています。

その他、フォロワー一覧に、記事投稿数、githubやtwitter連携のアイコン（リンク）などの情報があると良いかもしれません。スパム業者であるかどうかの判断だけではなく、フォローしてきたユーザーがどのような人であるかを知ることが出来ます。

**Comment by kskwtnk** [comment]
@ko1nksm 

ご意見ありがとうございます！

スパム問題は運営としてもずっと頭を抱えている問題でして……（今日のAdvent Calendarの件もそうです）

回答が若干ずれてしまうのですが、まず現在行っているスパムユーザーへの対策を共有しますと

- フォローアクションをとったユーザーのデータを用いたヒューリスティックなスパム判定
- 過去のスパムユーザーのデータを用いたスパム判定
    - ハックされてしまうとよろしくないので、ここでは伏せさせてください 🙇 

といったものになっています

ただしスパマーもどんどん手口が巧妙になっており、ko1nksmさんの仰るように「スパム行為をする意義が少なくなる」方向性に持っていきたいなとは思います

具体的にこれをします！と確約できませんが、常にスパムへの対策は考えていますので、今回のご意見も取り入れさせていただく可能性はあります 🙏

**Comment by ko1nksm** [comment]
状況が改善されている感じがしないので、再度お願いいたします。

ヒューリスティックなスパム判定など、技術的には面白いかと思いますが、現状うまくいっていないようです。もっと単純にキャプチャを導入して（最初に？）フォローする時に弾くとかできないものでしょうか？外国からのスパムが多いため、特に日本語を使うキャプチャであれば効果が高いように思えます。（すでにキャプチャを導入していたとしたらすみません）

通報が多いと、その対応で Qiita 様にも余計な作業コストが掛かっているのではないのではないかと思います。是非対応をよろしくお願いいたします。

**Reply by mkdk09** [reply]
@ko1nksm 
ご意見ありがとうございます。
キャプチャに関してはすでに導入しているのですが、それでも防げていない状況です...

対策が遅くなり、ご迷惑をおかけして申し訳ないのですが、社内でも問題意識を持って議論は行っているところです。
引き続き対策を検討していきますので、お待たせしてしまいますがご理解のほどお願いします。 :bow:

**Reply by ko1nksm** [reply]
承知しました。キャプチャはすでに導入されていましたか。うーん、まさか手動で登録とかしているのでしょうかね？それにしても外国の文字ばかりなので不自然な感じです。

私が何が気になっているのかを改めて考えてみたのですが、フォロワーのリンクをクリックした時に、そこが外国の文字で読めませんが広告だらけであることです。誰も広告を見たいとは思わないでしょう。

フォロワーのリンクをクリックするとき、それが自分のフォロワーであれ他人のフォロワーであれ、そこに期待する表示される情報は一体どんな人がフォローしているのだろうか？だと思います。フォロワーが少ない人は別として多い人の場合、そのフォロワーを全部見る人はいないでしょう。ほとんどが自分の知らない人だからです。つまりフォロワーのリンクをクリックした時に求めている情報は、最近フォローした人ではなく「ランクの高い人」だと思います。

つまりデフォルトで表示する情報は、最近フォローした順ではなく、ランクの高い人順で表示するのが良いのではないでしょうか？ランクが高い人の定義は投稿記事が多いとかフォロワーが多い人などが考えられると思います。スパムは必然的にランクが下るはずなので、目にしなくてすむのではないかと思います。

まあ、これはこのようにして欲しいという要望ではなく、ただの私の思いつきなので、参考程度として考えてください。

**Reply by shiracamus** [reply]
スパムフォロワーのプロフィールの三点メニューからブロックを選ぶとフォロワーから外すことができますよ。
面倒だけれど、私はそうしてきました。

**Reply by tanaka-asami415** [reply]
@ko1nksm 
ありがとうございます🙏
今回いただいた「ランクの高い人」順に表示されるようになると、
直近のフォロー順で見れなくなってしまうなどデメリットもありそうですが
フォロワーの見え方などは改善する必要があるため、スパムへの対応と併せて検討させていただきます🙇

**Reply by ko1nksm** [reply]
@shiracamus
はい。私も時々そのようにして消しているのですが、まあ、その、面倒なんですよね……。

@tanaka-asami415 
明確に書いていませんでしたが「デフォルトで表示する情報は、最近フォローした順ではなく」というのは、表示する順番を変更する機能をつけるということを意図していました。つまり必要な人は「直近のフォロー順で表示」をクリックするみたいな方法で切り替えられれば良いかなと。ただフォロワーの一覧をそこまで高機能にするのはメンテナンス性、ユーザービリティから考えて、私なら悩むところですね。

また、直近のフォローは自分のであれば通知一覧から知ることが出来るはずなので、直近のフォロー順での表示は不要な気もします。フォロワーの一覧を表示する機能は、一体誰がなんのために必要としているのだろうというのが考えるポイントな気がします。まあ、ここは仕様を煮詰める場所ではないと思うので、今後いつどのように対策なされるのかはおまかせします。

あと一つ思い出したのですが、どうもスパマーは登録が完了してからプロフィール内容を書き換えているようです。通知が来てすぐに内容が変更されたことを確認しました。お気づきかもしれませんが、それによってスパムフィルターをすり抜けているのかもしれません。

**Reply by tanaka-asami415** [reply]
@ko1nksm 
なるほど！詳細な情報をありがとうございます🙏

**Reply by ko1nksm** [reply]
調査してお気づきかもしれませんが、最近の状況として一日に10件以上スパムらしきフォローがあります。3日も放置すればフォロワー一覧の1ページ目がほぼスパムで埋まる状況です。

したがってユーザーランキングに載った人（おそらくここからフォローする人を決めているのでしょう）のフォロワー一覧機能は、事実上機能していないと思われます。

**Reply by tanaka-asami415** [reply]
@ko1nksm 
ご不便をおかけして、申し訳ありません🙇
スパム対応やフォロワーの見え方と合わせて、新規登録時のフローの見直しも検討しております。

多くのスパムにフォローされてしまう原因の一つとして
新規会員登録をされる方に、「おすすめユーザー」として表示されていることが考えられるためです。
（おすすめユーザーに表示されるユーザーは、直近のユーザーランキングも加味されているため、ユーザーランキングに入った方はフォロワー数が増えやすい傾向があります。）

時間はかかってしまうと思いますが、引き続き対策を検討していきますので、ご理解いただけますと幸いです🙏

**Comment by kyntk** [comment]
@ko1nksm
本日、メール未認証のユーザー、スパムと判定されたユーザーの一部の行動を制限しました。 :bow:
これによって、フォロワー一覧にも少し変化があると思われます。:pray:

**Reply by ko1nksm** [reply]
早い対応ありがとうございました。🙇
しばらく観察していたのですが、スパムらしきフォロワーの数は劇的に減ったように思えます。
（私がおすすめユーザーから消えただけかもしれませんが……）

それでもやはりスパムは来ているようです（ゼロにするのは不可能というのは承知しております）
プロフィールにURLなどをしっかり書いているものだけがすり抜けているような感じですね。

**Reply by kyntk** [reply]
@ko1nksm それは良かったです！

> それでもやはりスパムは来ているようです（ゼロにするのは不可能というのは承知しております）

はい、引き続き対策を検討していきます！
